### ISSUE TITLE
Variables used *before* assign should throw an error.

### ISSUE THREAD

**Issue by ConduciveMammal** [Shopify/theme-check#752]
**Describe the bug**

There's a check for undefined variables which works well, but all it seems to be doing is checking for the existence of an assignment, not the validity of the assignment.

Currently, I can attempt to access a variable before it's been assigned and it's considered valid.

If you look in ESLint, this would throw an undefined error.

**Expected**
```liquid
{{ cool }}
{% assign cool = true %}
// Variable "cool" used before being assigned.
```

**Actual**

```liquid
{{ cool }}
{% assign cool = true %}
// Totally cool 👍🏻 
```

**Debugging information**
- OS: macOS Sonoma
- Version: 1.15.0

![CleanShot 2023-12-13 at 4  50 33@2x](https://github.com/Shopify/theme-check/assets/18140157/e03971d1-d77d-4492-8622-ed51a8d685cc)
![CleanShot 2023-12-13 at 4  51 31@2x](https://github.com/Shopify/theme-check/assets/18140157/563c8289-93d9-4b13-b770-4402550e0f6e)

**Comment by lukeh-shopify** [comment]
👋🏻 Hi @ConduciveMammal! Thanks for reporting this. Unfortunately the Ruby version of theme check has been deprecated for some time. Would you be able to confirm you are still having issues when using our [maintained version of theme check](https://shopify.dev/docs/storefronts/themes/tools/theme-check)? This is done using either the latest version of the Shopify CLI, or via the Shopify Liquid Visual Studio Code extension.

If you are still having issues, please raise an issue over in the CLI or theme tools repositories. Thanks!
